src/validation.cpp

```diff
@@ -3649,7 +3661,23 @@ bool BlockManager::LoadBlockIndex(
         if (pindex->IsAssumedValid() ||
                 (pindex->IsValid(BLOCK_VALID_TRANSACTIONS) &&
                  (pindex->HaveTxsDownloaded() || pindex->pprev == nullptr))) {
-            block_index_candidates.insert(pindex);
+            // Fill each chainstate's block candidate set. Add all blocks as
+            // candidates if the chainstate is allowed to rely on assumed-valid
+            // blocks. Otherwise avoid adding assumed-valid blocks so the
+            // chainstate will download and validate all earlier blocks.
+            //
+            // Note: given the loop over `vSortedByHeight` and the way
+            // `seen_assumed_valid` works, it is (in general) possible that we
+            // are considering a block which has a height higher than the
+            // assumed-valid region but is not itself reliant on assumed-valid
+            // ancestors (i.e. a very deep fork). In practice this will not
+            // happen because assumed-valid blocks are buried deeply in
+            // the valid chain.
+            for (CChainState* chainstate : chainman.GetAll()) {
+                if (!seen_assumed_valid || chainstate->reliesOnAssumedValid()) {
+                    chainstate->setBlockIndexCandidates.insert(pindex);
```

Comment:
In commit "validation: have LoadBlockIndex account for snapshot use" (2833bd0277ac87cd07e98b0a23b73eed97a2e513)

I still think it would be really nice to call this function from a unit test and make sure this set is filled correctly. The logic here is fragile, and testing this part of `LoadBlockIndex` could make it easier to test other parts as well.

Reply:
Sorry for chiming in on what is now a very old PR, but I believe this block of code is incorrect. If we have a snapshot in place (and so some block index is marked IsAssumedValid), but we have downloaded blocks that haven't yet been activated (eg on the background-validation chainstate), this logic prevents us from adding those downloaded-but-not-yet verified blocks to `setBlockIndexCandidate` for the background chainstate. 

I ran into this while working on a download implementation for assumeutxo, and in some sequence of node restarts I got stuck in a state where my node had downloaded a bunch of blocks but never activated them, because they weren't in `setBlockIndexCandidates`.  I thought that restarting should fix the issue, but it didn't due to this block of code not doing the right thing...

Admittedly there is a lot of complicated interaction here but I think there's an invariant for `setBlockIndexCandidates` that is violated by this block of code, which is that if we have the data for a block and all its predecessors and it has more work than our tip, then it should be in `setBlockIndexCandidates`.  By excluding blocks that are marked IsAssumedValid() we violate this.

Reply:
@sdaftuar:

> If we have a snapshot in place (and so some block index is marked IsAssumedValid), but we have downloaded blocks that haven't yet been activated (eg on the background-validation chainstate), this logic prevents us from adding those downloaded-but-not-yet verified blocks to setBlockIndexCandidate for the background chainstate

I'm having a hard time following how this could happen. If the bg chainstate has some downloaded-but-not-activated blocks, they are definitionally beneath the `first_assumed_valid_height` (since we don't attach blocks above that height to the bg chainstate). So any of those blocks should be added by this code to the bg chainstate's `setBlockIndexCandidates`. Have you seen an example where you're sure that a downloaded block on the bg chain is excluded by this code?

I don't doubt that you're seeing an issue though... I wonder if there might be a problem with how `m_best_header` is set below? I.e. that state isn't qualified per chainstate. That's something I'll investigate now.

I'd also be curious if anyone else has seen this issue when testing the assumeutxo tracking branch (https://github.com/bitcoin/bitcoin/pull/27596). Is there a chance some code you've added is causing this?

> By excluding blocks that are marked IsAssumedValid() we violate this.

Note that we're only excluding blocks marked assumedvalid (by way of the height check) for the bg chainstate, which still seems sensible to me. All blocks, assumedvalid or not, are added to the snapshot chainstate (or "regular" chain if no snapshot is in use).

Reply:
> they are definitionally beneath the first_assumed_valid_height (since we don't attach blocks above that height to the bg chainstate)

I don't believe this is true; after we download a block and before we validate the block, the CBlockIndex entry for that block will have the AssumedValid bit set to true.  This means that if you were to download a bunch of blocks and then shutdown prior to those blocks being validated on the bg chainstate, that when you restart you'll be in this state.

Reply:
The code here isn't comparing with the snapshot base height; it's just checking if a block index entry is below the `first_assumed_valid_height`, and only in that case does it get added to `setBlockIndexCandidates`.  [I don't believe there's any other logic elsewhere that will do what you are describing, of adding entries below the snapshot base height to `setBlockIndexCandidates`?]

Reply:
More generally: I think that we don't need to worry at all about whether the assumed-valid bit is set when adding entries to `setBlockIndexCandidates`. We always validate a block when it gets connected, so the only invariant we need on `setBlockIndexCandidates` is whether they have enough work to be worth considering.  (We used to have a hard requirement that we always HAVE_DATA for such entries, but after pruning we had to relax that a bit, as there are edge-case scenarios where you might HAVE_DATA for such an entry and then prune it before you need it, so we handle that case by (re-)adding such entries to `mapBlocksUnlinked` when they occur. But I digress...)

Reply:
> The code here isn't comparing with the snapshot base height; it's just checking if a block index entry is below the first_assumed_valid_height

Oops, total misread on my part. You're right; `first_assumed_valid_height` will be the block after the tip of the background chain.

Reply:
The logic here does seem confusing.

Code comment says background candidate set needs to exclude BLOCK_ASSUMED_VALID blocks because otherwise ActivateBestChain "would add assumed-valid blocks to the chain". But that seems like exactly what it should do? (after validating them)

Commit message says background candidate set needs to exclude blocks after the snapshot height that don't have BLOCK_ASSUMED_VALID, but depend on blocks that do have BLOCK_ASSUMED_VALID. That idea does make sense, but does not seem to be what is implemented.

Taking another a step back, it seems like all of this would be simpler if background ActivateBestChain function just skipped calling FindMostWorkChain and tried to connect blocks up to the base of the snapshot.